---
title: "Class 15"
author: "Nadia Haghani"
date: "11/16/2021"
output:
  pdf_document: default
  html_document: default
---

# Background
Our data for tdoay comes from Himes et al. RNASeq analysis of the drug dexamethasone, a synthetic glucocorticoid steroid with anti-inflammatory effects (Himes et al 2014)

Read the countData and the colData.

```{r}
counts <- read.csv("airway_scaledcounts.csv", row.names=1)
metadata <- read.csv("airway_metadata.csv")
```

Let's have a look at these

```{r}
head(counts)
```


```{r}
metadata
```

> Q1. How many genes are in this dataset? 

```{r}
nrow(counts)
```


> Q2. How many ‘control’ cell lines do we have? 

```{r}
sum(metadata$dex == "control")
```

# Toy Differential gene expression


First I need to extract all of the "control" columns. Then, I will take the rowwise mean to get the average count values for all genes in these four experiments.

```{r}
control.inds <- metadata$dex == "control"
control.counts <- counts[ , control.inds]
head(control.counts)
```

```{r}
control.mean <- rowMeans(control.counts)
```


Now we do the same thing for the drug treated experiments (i.e. columns)
```{r}
treated.inds <- metadata$dex == "treated"
treated.counts <- counts[ , treated.inds]
head(treated.counts)
```

```{r}
treated.mean <- rowMeans(treated.counts)
```


Make a dataframe and add control with treated 

```{r}
meancounts <- data.frame(control.mean, treated.mean)
```

> Q5 (a). Create a scatter plot showing the mean of the treated samples against the mean of the control samples. Your plot should look something like the following.

Let's make a quick plot

```{r}
plot(meancounts)
```

This data needs a logarithmic transformation to better visualize all the data

```{r}
plot(meancounts, log ="xy")
```

We often use log2 in this field because it has nice math properties that make interpretation easier.

```{r}
log2(20/10)
log2(5/10)
```

We see 0 values for no change, + values for upregulations/increases, - values for downregulations/decreases. This nice property leads us to work with **log2(fold-change)** all the time in genomics and proteomics files.

Let's add the **log2(fold-change)** values to our 'meancounts' dataframe.

```{r}
meancounts$log2fc <- log2(meancounts[,"treated.mean"]/meancounts[,"control.mean"])
head(meancounts)
```

Exclude rows with zero values in our dataset. I can use the **which()** function with the 'arr.ind=TRUE' argument to get the columns and rows where the TRUE values are (ie the zero counts in our case)

```{r}
zero.vals <- which(meancounts[,1:2] == 0, arr.ind = TRUE)
head(zero.vals)
```
Take just the row
```{r}
to.rm <- unique(zero.vals[,"row"])
sort(to.rm)
```

Remove genes with zeros in rows

```{r}
mycounts <- meancounts[-to.rm,]
head(mycounts)
```

How many genes do we have left?

```{r}
nrow(mycounts)
```

How many genes were upregulated and downregulated?

```{r}
up.ind <- mycounts$log2fc > 2
down.ind <- mycounts$log2fc < (-2)
```

```{r}
sum(up.ind)
sum(down.ind)
```

So, there are 250 upregulated genes and 367 downregulated genes.


## DESeq2 Analysis

Let’s do this the right way. DESeq2 is an R package specifically for analyzing count-based NGS data like RNA-seq. It is available from Bioconductor.


```{r}
library(DESeq2)
# citation("DESeq2")
```

We need to first setup the input object for deseq2
```{r}
dds <- DESeqDataSetFromMatrix(countData=counts, 
                              colData=metadata, 
                              design=~dex)
dds
```

Now we can run DESeq analysis

```{r}
dds <- DESeq(dds)
```

To get at results, use the desequ 'results()' function:

```{r}
results(dds)
```

Run the DESeq pipeline on the dds object, and reassign the whole thing back to dds, which will now be a DESeqDataSet populated with all those values. 

```{r}
dds <- DESeq(dds)
```

Call the results() function on the DESeqDataSet that has been run through the pipeline.
```{r}
res <- results(dds)
res
```

## Volcano Plots

Let’s make a commonly produced visualization from this data, namely a so-called Volcano plot. These summary figures are frequently used to highlight the proportion of genes that are both significantly regulated and display a high fold change.


```{r}
plot(res$log2FoldChange, log(res$padj))
```

We can flip the pvalue axis by adding a '-' sign before the log. 
```{r}
plot(res$log2FoldChange,  -log(res$padj), 
      xlab="Log2(FoldChange)",
      ylab="-Log(P-value)")
```


This plot is not a useful plot because the small p-values are hidden at the bottom. We can make the plot more useful by adding color, cut-off lines, highlighting genes with high fold changes...

```{r}
# Setup our custom point color vector 
mycols <- rep("gray", nrow(res))
mycols[ abs(res$log2FoldChange) > 2 ]  <- "red" 

inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

# Volcano plot with custom colors 
plot( res$log2FoldChange,  -log(res$padj), 
 col=mycols, ylab="-Log(P-value)", xlab="Log2(FoldChange)" )

# Cut-off lines
abline(v=c(-2,2), col="gray", lty=2)
abline(h=-log(0.1), col="gray", lty=2)

```




